test/embedding/embeddingdl.c

```diff
@@ -0,0 +1,198 @@
+    const char *library_name = "libjulia.dylib";
+#else
+    const char *library_name = "libjulia.so";
+#endif
+    void *libjulia = dlopen(library_name, RTLD_LAZY | RTLD_GLOBAL);
+#endif
+
+    if (!libjulia) {
+        fprintf(stderr, "Failed to load libjulia.\n");
+        return 0;
+    }
+
+    int success = 1;
+    p_jl_init = load_function(libjulia, "jl_init", &success);
+    p_jl_is_initialized = load_function(libjulia, "jl_is_initialized", &success);
+    p_jl_atexit_hook = load_function(libjulia, "jl_atexit_hook", &success);
+    p_jl_eval_string = load_function(libjulia, "jl_eval_string", &success);
+    p_jl_symbol = load_function(libjulia, "jl_symbol", &success);
+    p_jl_main_module = load_function(libjulia, "jl_main_module", &success);
+    p_jl_get_global = load_function(libjulia, "jl_get_global", &success);
+    p_jl_unbox_voidpointer = load_function(libjulia, "jl_unbox_voidpointer", &success);
+    p_jl_string_ptr = load_function(libjulia, "jl_string_ptr", &success);
+    p_jl_exception_occurred = load_function(libjulia, "jl_exception_occurred", &success);
+
+    return success;
+}
+
+// Masquerade the dynamically loaded function pointers as regular functions.
+#define jl_init (*p_jl_init)
+#define jl_is_initialized (*p_jl_is_initialized)
+#define jl_atexit_hook (*p_jl_atexit_hook)
+#define jl_eval_string (*p_jl_eval_string)
+#define jl_symbol (*p_jl_symbol)
+#define jl_main_module (*p_jl_main_module)
+#define jl_get_global (*p_jl_get_global)
+#define jl_unbox_voidpointer (*p_jl_unbox_voidpointer)
+#define jl_string_ptr (*p_jl_string_ptr)
+#define jl_exception_occurred (*p_jl_exception_occurred)
+
+// Helper function to retrieve pointers to cfunctions on the Julia side.
+static void *get_cfunction_pointer(const char *name)
+{
+    void *p = 0;
+    jl_value_t *boxed_pointer = jl_get_global(jl_main_module, jl_symbol(name));
+
+    if (boxed_pointer != 0) {
+        p = jl_unbox_voidpointer(boxed_pointer);
+    }
+
+    if (!p) {
+        fprintf(stderr, "cfunction pointer %s not available.\n", name);
+    }
+
+    return p;
+}
+
+// jl_eval_string with error checking.
+static int checked_eval_string(const char *command, const char *error_message)
+{
+    jl_eval_string(command);
+
+    if (jl_exception_occurred()) {
+        const char *p = jl_string_ptr(jl_eval_string("sprint(showerror, ccall(:jl_exception_occurred, Any, ()))"));
+        fprintf(stderr, "%s%s\n", error_message, p);
+        return 0;
+    }
+
+    return 1;
+}
+
+int main()
+{
+    if (!load_julia_functions())
+        return EXIT_FAILURE;
+
+    jl_init();
+    if (!jl_is_initialized()) {
+        fprintf(stderr, "jl_init failed.");
+        jl_atexit_hook(EXIT_FAILURE);
+        return EXIT_FAILURE;
+    }
+
+    // `include` is not available out of the box from the embedding
+    // environment, so we add it here (issue #28825).
+    checked_eval_string("include(x) = Base.include(Main, x)",
+                        "Failed to define include: ");
```

Comment:
Will be fixed in #32062, but still required for anyone targeting 1.0, unless that's backported.

Reply:
This is now merged, so you should be able to remove this.